File: bebek.cpp
#include <iostream>
#include <string>
#include "FarmAnimal.h"
#include "Eggproducing.h"
#include "Meatproducing.h"
#include "bebek.h"

using namespace std;

const int a = 4;
const int b =4;
const int c = 4;
const int d =4;

bebek :: bebek(int posX,int posY ) : Eggproducing(posX,posY),Meatproducing(posX,posY),FarmAnimal(posX,posY){
    Full =FULLMAX;
    countTelurB =0;
    this->posX = FarmAnimal::getPosX();
    this->posY = FarmAnimal::getPosY();

}
bebek:: ~bebek(){
    this->posX =0;
    this->posY =0;
    Full =0;
    countTelurB=0;
    //cout<<"One of your duck has died"<<endl;
    //FarmAnimal::~FarmAnimal();
}
void bebek::move (){
   int x = rand()%(a+b-1);
    int y = rand()%(a+b-1);
    //cout<<" x , y "<<x<< " " <<y<<endl;
    this->posX+=x;
    this->posY +=y;
    while(this->posX>a){
         x = rand()%(a+b-1);
        //cout<<" x "<<x<<endl;
        while (this->posY>b){
            this->posY -=x;
        }
        this->posX -=x;
        if (this->posX <a){
            this->posX = this->posX + a;
        }
    }
    
    Full--;
}
void bebek::talk(){
    cout<<"Qwek ... qwek"<<endl;

}
void bebek::eat(){
    countTelurB++;
    Full=FULLMAX;
     if (countTelurB > MAX) {
	countTelurB = MAX;
    }

}
string bebek::getProduct(){};
//void setProduct(string telur);
int bebek::getcountTelurB()const{
    return countTelurB;
}
int bebek::getFull(){
    return Full;
}
void bebek::setFull(int Full){
    this->Full = Full;
}
void bebek::Print(){
    cout<<"Telur b ada " <<countTelurB<<endl;
    cout<<"Full : "<< Full<<endl;
    cout<<"pos : "<< this->posX<<" " << this->posY<<endl;
}
/*void bebek::setPosX(int x){
    posX = x;
}
void bebek::setPosY(int y){
    posX = x;
}
int bebek ::getPosX(){
    return posX;
}
int bebek ::getPosY(){
    return posY;
}*/

File: test_GUI/Bebek.java

import java.util.*;
import java.util.ArrayList;
import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.awt.image.BufferStrategy;
import java.io.IOException;
import java.util.concurrent.ThreadLocalRandom;

public class Bebek extends Farm_Animal {
    private BufferedImage[] sprites;
    private int countTelurB;
    private boolean isTalk;

    private Graphics2D g;

    private BufferedImage image;

    public Bebek(FarmMap fm){
        super(fm);

        moveSpeed =1;
        maxSpeed =10;

        width =30;
        height =30;
        cwidth =20;
        cheight =20;

        isTalk = false;
        // load sprite
        try{
            BufferedImage sheet = ImageIO.read(getClass().getResourceAsStream("duck.png"));
            sprites = new BufferedImage[3];
            for(int i =0;i<sprites.length;i++){
                sprites[i] = sheet.getSubimage(1*width,0,width,height);
            }

        } catch (Exception e){
            e.printStackTrace();
        }
        animation = new Animation();
        animation.setFrames(sprites);
        animation.setDelay(300);

        left = true;

    }

    public void move(){
        int x = ThreadLocalRandom.current().nextInt(-randX, randY + 1);
        int y =ThreadLocalRandom.current().nextInt(-randX, randY);
        if(!nothidup()){
            if(getX()+x>30 &&getX()+x<GamePanel.w-30&&getY()+y>10&&getY()+y<GamePanel.h-30){
                //dx += moveSpeed;
                setPosition(getX()+x,getY()+x);
            }
            // } else {
            //     setPosition(getX()-x*2,getY()-x*2);
            // }
            setFull(getFull()-0.01); 
        }
        


    }
    public void eat(Graphics2D g){
        if (getFull() ==HUNGRY){
            setFull(FULLMAX);
            countTelurB =MAX;    
        }
        //super.draw(g);

        //animation.update();


    }

    public String getChar(){
        return "D";
    }

    public void talk(){
        System.out.println(DUCK);
    }

    public String getTalk(){
        return DUCK;
    }

    public boolean isTalk() {
        return isTalk;
    }

    public void setIsTalk(boolean val) {
        isTalk = val;
    }


    public void getProduct(){
        this.countTelurB =0;
    }

    public String nameProduct(){
        return "DuckMeat";
    }
    public String getEggandMilk(){
        return "DuckEgg";
    }

    public void update(){
        move();
        //talk();
        //Print();
        eat(g);
        checkFarmMapCollision();
        setPosition(xtemp,ytemp);
        animation.update();
    }

    public void draw(Graphics2D g){
        if(notOnScreen()) return;
        setMapPosition();
        // if (eat){
        //     g.setRGB(getX(),getY(),30,30,Color.WHITE);
        // }
        super.draw(g);
    }

    public void Print(){
        System.out.println("Bebek " + countTelurB + " Full : "+ getFull()+" pos :  "+ getX() +" "+getY());
    }
}